manul/src/combinators/chain.rs

```diff
@@ -229,118 +225,138 @@ where
-    C: Chained<Id>,
+    C: ChainedProtocol,
 {
+    type Result = <C::Protocol2 as Protocol>::Result;
+    type ProtocolError = ChainedProtocolError<C>;
+    type CorrectnessProof = ChainedCorrectnessProof<C>;
 }
 
-/// The protocol resulting from chaining two sub-protocols as described by `C`.
-#[derive(Debug)]
-#[allow(clippy::type_complexity)]
-pub struct ChainedProtocol<Id: PartyId, C: Chained<Id>>(PhantomData<fn((Id, C)) -> (Id, C)>);
+/// A trait defining how the entry point for the whole chained protocol
+/// will be split into the entry point for the first protocol, and a piece of data
+/// that, along with the first protocol's result, will be used to create the entry point for the second protocol.
+pub trait ChainedSplit<Id: PartyId> {
+    /// The chained protocol this trait belongs to.
+    type Protocol: ChainedProtocol;
 
-impl<Id, C> Protocol for ChainedProtocol<Id, C>
-where
-    Id: PartyId,
-    C: Chained<Id>,
-{
-    type Result = <<C::EntryPoint2 as EntryPoint<Id>>::Protocol as Protocol>::Result;
-    type ProtocolError = ChainedProtocolError<Id, C>;
-    type CorrectnessProof = ChainedCorrectnessProof<Id, C>;
-}
+    /// The first protocol's entry point.
+    type EntryPoint: EntryPoint<Id, Protocol = <Self::Protocol as ChainedProtocol>::Protocol1>;
 
-/// The entry point of the chained protocol.
-#[derive_where::derive_where(Debug)]
-pub struct ChainedEntryPoint<Id: PartyId, C: Chained<Id>> {
-    state: ChainState<Id, C>,
+    /// Creates the first protocol's entry point and the data for creating the second entry point.
+    fn make_entry_point1(self) -> (Self::EntryPoint, impl ChainedJoin<Id, Protocol = Self::Protocol>);
 }
 
-#[derive_where::derive_where(Debug)]
-enum ChainState<Id, C>
-where
-    Id: PartyId,
-    C: Chained<Id>,
-{
-    Protocol1 {
-        round: BoxedRound<Id, <C::EntryPoint1 as EntryPoint<Id>>::Protocol>,
-        shared_randomness: Box<[u8]>,
-        id: Id,
-        inputs: C::Inputs,
-    },
-    Protocol2(BoxedRound<Id, <C::EntryPoint2 as EntryPoint<Id>>::Protocol>),
+/// A trait defining how the data created in [`ChainedSplit::make_entry_point1`]
+/// will be joined with the result of the first protocol to create an entry point for the second protocol.
+pub trait ChainedJoin<Id: PartyId>: 'static + Debug + Send + Sync {
+    /// The chained protocol this trait belongs to.
+    type Protocol: ChainedProtocol;
+
+    /// The second protocol's entry point.
+    type EntryPoint: EntryPoint<Id, Protocol = <Self::Protocol as ChainedProtocol>::Protocol2>;
+
+    /// Creates the second protocol's entry point using the first protocol's result.
+    fn make_entry_point2(
+        self,
+        result: <<Self::Protocol as ChainedProtocol>::Protocol1 as Protocol>::Result,
+    ) -> Self::EntryPoint;
 }
 
-impl<Id, C> EntryPoint<Id> for ChainedEntryPoint<Id, C>
+impl<Id, T> EntryPoint<Id> for T
 where
     Id: PartyId,
-    C: Chained<Id>,
+    T: ChainedSplit<Id> + CombinatorEntryPoint<Type = Chain>,
```

Comment:
```suggestion
    T: ChainedSplit<Id> + CombinatorEntryPoint<Kind = Chain>,
```

I think it'd be good to avoid the word "type" when we can.